timeutil/schedule.go

```diff
@@ -28,49 +29,57 @@ import (
 	"time"
 )
 
-var validTime = regexp.MustCompile(`^([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9])$`)
+// Match 0:00-24:00, where 24:00 means the later end of the day.
+var validTime = regexp.MustCompile(`^([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9])$|^24:00$`)
 
-type TimeOfDay struct {
+// Clock represents a hour:minute time within a day.
+type Clock struct {
 	Hour   int
 	Minute int
 }
 
-func (t TimeOfDay) String() string {
+func (t Clock) String() string {
 	return fmt.Sprintf("%02d:%02d", t.Hour, t.Minute)
 }
 
-// Sub subtracts `other` TimeOfDay from current and returns duration
-func (t TimeOfDay) Sub(other TimeOfDay) time.Duration {
+// Sub returns the duration t - other.
+func (t Clock) Sub(other Clock) time.Duration {
 	t1 := time.Duration(t.Hour)*time.Hour + time.Duration(t.Minute)*time.Minute
 	t2 := time.Duration(other.Hour)*time.Hour + time.Duration(other.Minute)*time.Minute
 	return t1 - t2
 }
 
-// Add adds given duration and returns a new TimeOfDay
-func (t TimeOfDay) Add(dur time.Duration) TimeOfDay {
+// Add adds given duration to t and returns a new Clock
+func (t Clock) Add(dur time.Duration) Clock {
 	t1 := time.Duration(t.Hour)*time.Hour + time.Duration(t.Minute)*time.Minute
 	t2 := t1 + dur
-	nt := TimeOfDay{
+	nt := Clock{
 		Hour:   int(t2.Hours()) % 24,
 		Minute: int(t2.Minutes()) % 60,
 	}
 	return nt
 }
 
-// IsValidWeekday returns true if given s looks like a valid weekday. Valid
-// inputs are 3 letter, lowercase abbreviations of week days.
-func IsValidWeekday(s string) bool {
-	_, ok := weekdayMap[s]
-	return ok
+// Time generates a time.Time with hour and minute set from t, while year, month
+// and day are taken from base
+func (t Clock) Time(base time.Time) time.Time {
+	return time.Date(base.Year(), base.Month(), base.Day(),
+		t.Hour, t.Minute, 0, 0, base.Location())
 }
 
-// ParseTime parses a string that contains hour:minute and returns
-// an TimeOfDay type or an error
-func ParseTime(s string) (t TimeOfDay, err error) {
+// ParseClock parses a string that contains hour:minute and returns
+// a Clock type or an error
+func ParseClock(s string) (t Clock, err error) {
 	m := validTime.FindStringSubmatch(s)
 	if len(m) == 0 {
 		return t, fmt.Errorf("cannot parse %q", s)
 	}
+
+	if m[0] == "24:00" {
+		t.Hour = 24
```

Comment:
Don't you need to set t.Minute as well? Also why is this not just handled generally below?

Reply:
`t.Minutes` will default to 0 which is just fine in this case.